advanced_source/ddp_pipeline_tutorial.py

```diff
@@ -0,0 +1,452 @@
+# length 6:
+#
+# .. math::
+#   \begin{bmatrix}
+#   \text{A} & \text{B} & \text{C} & \ldots & \text{X} & \text{Y} & \text{Z}
+#   \end{bmatrix}
+#   \Rightarrow
+#   \begin{bmatrix}
+#   \begin{bmatrix}\text{A} \\ \text{B} \\ \text{C} \\ \text{D} \\ \text{E} \\ \text{F}\end{bmatrix} &
+#   \begin{bmatrix}\text{G} \\ \text{H} \\ \text{I} \\ \text{J} \\ \text{K} \\ \text{L}\end{bmatrix} &
+#   \begin{bmatrix}\text{M} \\ \text{N} \\ \text{O} \\ \text{P} \\ \text{Q} \\ \text{R}\end{bmatrix} &
+#   \begin{bmatrix}\text{S} \\ \text{T} \\ \text{U} \\ \text{V} \\ \text{W} \\ \text{X}\end{bmatrix}
+#   \end{bmatrix}
+#
+# These columns are treated as independent by the model, which means that
+# the dependence of ``G`` and ``F`` can not be learned, but allows more
+# efficient batch processing.
+#
+
+# In 'run_worker'
+    def print_with_rank(msg):
+        print('[RANK {}]: {}'.format(rank, msg))
+
+    import torchtext
+    from torchtext.data.utils import get_tokenizer
+    TEXT = torchtext.data.Field(tokenize=get_tokenizer("basic_english"),
+                                init_token='<sos>',
+                                eos_token='<eos>',
+                                lower=True)
+    train_txt, val_txt, test_txt = torchtext.datasets.WikiText2.splits(TEXT)
+    TEXT.build_vocab(train_txt)
+    device = torch.device(2 * rank)
+
+    def batchify(data, bsz, rank, world_size, is_train=False):
+        data = TEXT.numericalize([data.examples[0].text])
+        # Divide the dataset into bsz parts.
+        nbatch = data.size(0) // bsz
+        # Trim off any extra elements that wouldn't cleanly fit (remainders).
+        data = data.narrow(0, 0, nbatch * bsz)
+        # Evenly divide the data across the bsz batches.
+        data = data.view(bsz, -1).t().contiguous()
+        # Divide the data across the ranks only for training data.
+        if is_train:
+            data_per_rank = data.size(0) // world_size
+            data = data[rank * data_per_rank : (rank + 1) * data_per_rank]
+        return data.to(device)
+
+    batch_size = 20
+    eval_batch_size = 10
+    train_data = batchify(train_txt, batch_size, rank, world_size, True)
+    val_data = batchify(val_txt, eval_batch_size, rank, world_size)
+    test_data = batchify(test_txt, eval_batch_size, rank, world_size)
+
+
+######################################################################
+# Functions to generate input and target sequence
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+
+
+######################################################################
+# ``get_batch()`` function generates the input and target sequence for
+# the transformer model. It subdivides the source data into chunks of
+# length ``bptt``. For the language modeling task, the model needs the
+# following words as ``Target``. For example, with a ``bptt`` value of 2,
+# we’d get the following two Variables for ``i`` = 0:
+#
+# .. image:: ../_static/img/transformer_input_target.png
```

Comment:
how does this line help? github does not show the picture inline

Reply:
You can view the tutorial in a browser: https://deploy-preview-1347--pytorch-tutorials-preview.netlify.app/advanced/ddp_pipeline_tutorial.html.